cpp/src/arrow/engine/substrait/options.cc

```diff
@@ -122,11 +122,23 @@ class DefaultExtensionProvider : public BaseExtensionProvider {
   }
 };
 
-std::shared_ptr<ExtensionProvider> ExtensionProvider::kDefaultExtensionProvider =
+namespace {
+
+std::shared_ptr<ExtensionProvider> kDefaultExtensionProvider =
     std::make_shared<DefaultExtensionProvider>();
 
+std::mutex kDefaultExtensionProviderMutex;
```

Comment:
```suggestion
std::shared_ptr<ExtensionProvider> default_extension_provider =
    std::make_shared<DefaultExtensionProvider>();

std::mutex default_extension_provider_mutex;
```

These aren't constants so we can't use `kName`.  Will this work?

Reply:
You're right we can't use `kName`. I believe the right way to use is `g_name`, which I just used in the recent commit.